File: AlgorithmsPackPythonSolutions/Number of Closed Islands.py
class Solution:
    def closedIsland(self, grid: List[List[int]]) -> int:
        dr = [-1, 0, 1, 0]
        dc = [0, 1, 0, -1]

        def isValid(r: int, c: int, grid: List[List[int]]) -> bool:
            if r < 0 or r >= len(grid):
                return False
            if c < 0 or c >= len(grid[0]):
                return False
            return True

        def onBoundary(r: int, c: int, grid: List[List[int]]) -> bool:
            if r == 0 or r == len(grid) - 1:
                return True
            if c == 0 or c == len(grid[0]) - 1:
                return True
            return False

        def dfs(r: int, c: int, grid: List[List[int]], CCIDS: List[List[bool]], ccid: int) -> None:
            nonlocal touches_boundary
            if not isValid(r, c, grid) or CCIDS[r][c] or grid[r][c] == 1:
                return

            if onBoundary(r, c, grid):
                touches_boundary = True

            CCIDS[r][c] = ccid

            for d in range(4):
                dfs(r + dr[d], c + dc[d], grid, CCIDS, ccid)

        rows = len(grid)
        cols = len(grid[0])
        connectedComponentIDS = [[False] * cols for _ in range(rows)]
        ccid = 0
        count = 0

        for r in range(rows):
            for c in range(cols):
                if not grid[r][c] and not connectedComponentIDS[r][c]:
                    touches_boundary = False
                    dfs(r, c, grid, connectedComponentIDS, ccid + 1)
                    count += not touches_boundary

        return count

File: AlgorithmsPackJavaSolutions/Number of Closed Islands.java
class Solution {
    private final int[] dr = {-1, 0, 1, 0};
    private final int[] dc = {0, 1, 0, -1};

    private boolean touchesBoundary;

    private boolean isValid(int r, int c, int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        return r >= 0 && r < rows && c >= 0 && c < cols;
    }

    private boolean onBoundary(int r, int c, int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        return r == 0 || r == rows - 1 || c == 0 || c == cols - 1;
    }

    private void dfs(int r, int c, int[][] grid, boolean[][] connectedComponentIDS, int ccid) {
        if (!isValid(r, c, grid) || connectedComponentIDS[r][c] || grid[r][c] == 1) {
            return;
        }

        if (onBoundary(r, c, grid)) {
            touchesBoundary = true;
        }

        connectedComponentIDS[r][c] = true;

        for (int d = 0; d < 4; d++) {
            dfs(r + dr[d], c + dc[d], grid, connectedComponentIDS, ccid);
        }
    }

    public int closedIsland(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        boolean[][] connectedComponentIDS = new boolean[rows][cols];
        int ccid = 0;
        int count = 0;

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == 0 && !connectedComponentIDS[r][c]) {
                    touchesBoundary = false;
                    dfs(r, c, grid, connectedComponentIDS, ++ccid);
                    if (!touchesBoundary) {
                        count++;
                    }
                }
            }
        }

        return count;
    }
}
